braindecode/models/functions_attention.py

```diff
@@ -0,0 +1,467 @@
+        a = self.smoothing.unsqueeze(0).unsqueeze(0) * (x_ - cw_).pow(2).sum(3)
+        a = torch.softmax(a, dim=2)  # b x t x k
+
+        # aggregate
+        e = (a.unsqueeze(3) * (x_ - cw_)).sum(1)  # b x k x c
+        e_norm = torch.relu(self.bn(e)).mean(1)  # b x c
+
+        scale = torch.sigmoid(self.fc(e_norm))
+        return x * scale.unsqueeze(2).unsqueeze(3)
+
+
+class ECA(nn.Module):
+    """
+    XXX. TODO: Add description.
+    """
+    def __init__(self, in_channels: int, kernel_size: int):
+        super(ECA, self).__init__()
+        self.gap = nn.AdaptiveAvgPool2d(1)
+        assert kernel_size % 2 == 1, "kernel size must be odd for same padding"
+        self.conv = nn.Conv1d(1, 1, kernel_size=kernel_size, padding=kernel_size // 2,
+                              bias=False)
+
+    def forward(self, x):
+        scale = self.gap(x)
+        scale = rearrange(scale, "b c 1 1 -> b 1 c")
+        scale = self.conv(scale)
+        scale = torch.sigmoid(rearrange(scale, "b 1 c -> b c 1 1"))
+        return x * scale
+
+
+class GatherExcite(nn.Module):
+    """
+    XXX. TODO: Add description.
+    """
+    def __init__(self, in_channels: int, seq_len: int = 62, extra_params: bool = False,
+                 use_mlp: bool = False, reduction_rate: int = 4):
+        super(GatherExcite, self).__init__()
+        if extra_params:
+            self.gather = nn.Sequential(
+                nn.Conv2d(in_channels, in_channels, (1, seq_len), groups=in_channels,
+                          bias=False),
+                nn.BatchNorm2d(in_channels)
+            )
+        else:
+            self.gather = nn.AdaptiveAvgPool2d(1)
+
+
+        if use_mlp:
+            self.mlp = nn.Sequential(
+                nn.Conv2d(in_channels, int(in_channels // reduction_rate), 1,
+                          bias=False),
+                nn.ReLU(),
+                nn.Conv2d(int(in_channels // reduction_rate), in_channels, 1,
+                          bias=False),
+            )
+        else:
+            self.mlp = nn.Identity()
+
+    def forward(self, x):
+        scale = self.gather(x)
+        scale = torch.sigmoid(self.mlp(scale))
+        return scale * x
+
+
+class GCT(nn.Module):
+    """
+    XXX. TODO: Add description.
+    """
+    def __init__(self, in_channels: int):
+        super(GCT, self).__init__()
+        self.alpha = nn.Parameter(torch.ones(1, in_channels, 1, 1))
+        self.beta = nn.Parameter(torch.zeros(1, in_channels, 1, 1))
+        self.gamma = nn.Parameter(torch.zeros(1, in_channels, 1, 1))
+
+    def forward(self, x, eps: float = 1e-5):
+        embedding = (x.pow(2).sum((2, 3), keepdim=True) + eps).pow(0.5) * self.alpha
+        norm = self.gamma / (embedding.pow(2).mean(dim=1, keepdim=True) + eps).pow(0.5)
+        gate = 1.0 + torch.tanh(embedding * norm + self.beta)
+        return x * gate
+
+
+class SRM(nn.Module):
+    """
+    XXX. TODO: Add description.
+    """
```

Comment:
```suggestion
    """
    Parameters
    ----------
    in_channels : int, number of input feature channels
    use_mlp : bool, whether to use fully-connected layers instead of a convolutional layer, default=False
    reduction_rate : int, reduction ratio of the fully-connected layers (if used), default=4
    
    References
    ----------
    .. [Lee2019] Lee, H., Kim, H., Nam, H., 2019.
    SRM : A Style-based Recalibration Module for Convolutional Neural Networks. ICCV 2019.
    """
```